From: "K. Y. Srinivasan" <kys@microsoft.com>
Date: Mon, 7 Jul 2014 16:34:25 -0700
Patch-mainline: v3.16-rc5
Subject: Drivers: hv: util: Fix a bug in the KVP code
Git-commit: 9bd2d0dfe4714dd5d7c09a93a5c9ea9e14ceb3fc
References: bnc#886840

Add code to poll the channel since we process only one message
at a time and the host may not interrupt us. Also increase the
receive buffer size since some KVP messages are close to 8K bytes in size.

Content tweaked as per bnc above.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Cc: <stable@vger.kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Acked-by: <ohering@suse.de>
---
 drivers/hv/hv_kvp.c  |    4 ++--
 drivers/hv/hv_util.c |    2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

Index: linux-3.0-SLE11-SP3/drivers/hv/hv_kvp.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/drivers/hv/hv_kvp.c
+++ linux-3.0-SLE11-SP3/drivers/hv/hv_kvp.c
@@ -568,6 +568,7 @@ response_done:
 	vmbus_sendpacket(channel, recv_buffer, buf_len, req_id,
 				VM_PKT_DATA_INBAND, 0);
 
+	hv_kvp_onchannelcallback(channel);
 }
 
 /*
@@ -603,7 +603,7 @@ void hv_kvp_onchannelcallback(void *cont
 		return;
 	}
 
-	vmbus_recvpacket(channel, recv_buffer, PAGE_SIZE * 2, &recvlen,
+	vmbus_recvpacket(channel, recv_buffer, PAGE_SIZE * 4, &recvlen,
 			 &requestid);
 
 	if (recvlen > 0) {
Index: linux-3.0-SLE11-SP3/drivers/hv/hv_util.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/drivers/hv/hv_util.c
+++ linux-3.0-SLE11-SP3/drivers/hv/hv_util.c
@@ -312,7 +312,7 @@ static int util_probe(struct hv_device *
 		(struct hv_util_service *)dev_id->driver_data;
 	int ret;
 
-	srv->recv_buffer = kmalloc(PAGE_SIZE * 2, GFP_KERNEL);
+	srv->recv_buffer = kmalloc(PAGE_SIZE * 4, GFP_KERNEL);
 	if (!srv->recv_buffer)
 		return -ENOMEM;
 	if (srv->util_init) {
